core/src/main/java/org/apache/iceberg/SnapshotScan.java

```diff
@@ -0,0 +1,166 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.iceberg.events.Listeners;
+import org.apache.iceberg.events.ScanEvent;
+import org.apache.iceberg.expressions.ExpressionUtil;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.metrics.DefaultMetricsContext;
+import org.apache.iceberg.metrics.ImmutableScanReport;
+import org.apache.iceberg.metrics.ScanMetrics;
+import org.apache.iceberg.metrics.ScanMetricsResult;
+import org.apache.iceberg.metrics.ScanReport;
+import org.apache.iceberg.metrics.Timer;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.TypeUtil;
+import org.apache.iceberg.util.DateTimeUtil;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is a common base class to share code between different BaseScan implementations that handle
+ * scans of a particular snapshot.
+ *
+ * @param <ThisT> actual BaseScan implementation class type
+ * @param <T> type of ScanTask returned
+ * @param <G> type of ScanTaskGroup returned
+ */
+public abstract class SnapshotScan<ThisT, T extends ScanTask, G extends ScanTaskGroup<T>>
+    extends BaseScan<ThisT, T, G> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotScan.class);
+
+  private ScanMetrics scanMetrics;
+
+  protected SnapshotScan(Table table, Schema schema, TableScanContext context) {
+    super(table, schema, context);
+  }
+
+  protected Long snapshotId() {
+    return context().snapshotId();
+  }
+
+  protected abstract CloseableIterable<T> doPlanFiles();
+
+  protected ScanMetrics scanMetrics() {
+    if (scanMetrics == null) {
+      this.scanMetrics = ScanMetrics.of(new DefaultMetricsContext());
+    }
+
+    return scanMetrics;
+  }
+
+  @Override
+  public Table table() {
+    return super.table();
+  }
+
+  public ThisT useSnapshot(long scanSnapshotId) {
+    Preconditions.checkArgument(
+        snapshotId() == null, "Cannot override snapshot, already set snapshot id=%s", snapshotId());
+    Preconditions.checkArgument(
+        table().snapshot(scanSnapshotId) != null,
+        "Cannot find snapshot with ID %s",
+        scanSnapshotId);
+    TableScanContext newContext = context().useSnapshotId(scanSnapshotId);
```

Comment:
nit: Redundant variable? I think the call below fits on one line now so we can get rid of this var. You could use the temp var too, up to you. Just make it consistent here and below.

Reply:
Yea, thanks, I think this may have came from a bad rebase, reverted.